rest/rest-resources/src/main/java/org/apache/brooklyn/rest/filter/LoggingResourceFilter.java

```diff
@@ -0,0 +1,235 @@
+
+    /** Methods logged at trace. */
+    private static final Set<String> UNINTERESTING_METHODS = ImmutableSet.of("GET", "HEAD");
+
+    /** Headers whose values will not be logged. */
+    private static final Set<String> CENSORED_HEADERS = ImmutableSet.of("Authorization", "Cookie");
+
+    /** Log all requests that take this time or longer to complete. */
+    private static final Duration REQUEST_DURATION_LOG_POINT = Duration.FIVE_SECONDS;
+
+    /**
+     * Used to correlate the requests and responses - records the timestamp (in nanos) when the  
+     * request was received, so can calculate how long it took to process the request.
+     */
+    private final WeakHashMap<HttpServletRequest, Long> requestTimestamps = new WeakHashMap<>();
+    
+    @Context
+    private HttpServletRequest servletRequest;
+
+    public LoggingResourceFilter() {
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext) throws IOException {
+        requestTimestamps.put(servletRequest, System.nanoTime());
+        
+        String method = requestContext.getMethod();
+        boolean isInteresting = !UNINTERESTING_METHODS.contains(method.toUpperCase());
+        LogLevel logLevel = isInteresting ? LogLevel.DEBUG : LogLevel.TRACE;
+        
+        logRequest(requestContext, logLevel);
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) throws IOException {
+        Long requestTime = requestTimestamps.remove(servletRequest);
+        Duration requestDuration = null;
+        if (requestTime != null) {
+            long responseTime = System.nanoTime();
+            requestDuration = Duration.nanos(responseTime - requestTime);
+        }
+        
+        String method = requestContext.getMethod();
+        boolean isInteresting = !UNINTERESTING_METHODS.contains(method.toUpperCase())
+                || (requestDuration != null && requestDuration.isLongerThan(REQUEST_DURATION_LOG_POINT));
+        LogLevel logLevel = isInteresting ? LogLevel.DEBUG : LogLevel.TRACE;
+
+        logResponse(requestContext, responseContext, requestDuration, logLevel);
+    }
+
+    private void logRequest(ContainerRequestContext requestContext, LogLevel level) {
+        if (!isLogEnabled(LOG, level)) return;
+        
+        String method = requestContext.getMethod();
+        String path = requestContext.getUriInfo().getPath();
+        requestContext.getSecurityContext();
+        
+        SecurityContext securityContext = requestContext.getSecurityContext();
+        Principal userPrincipal = (securityContext != null) ? requestContext.getSecurityContext().getUserPrincipal() : null;
+        String userName = (userPrincipal != null) ? userPrincipal.getName() : "<no-user>";
+        String remoteAddr = servletRequest.getRemoteAddr();
+        
+        StringBuilder message = new StringBuilder("Request received: ")
```

Comment:
Would it be worth adding a correlation value here and in the response, so the two can be easily related to each other?  Maybe just the `hashCode()` of the `servletRequest` object?

Reply:
@geomacy @aledsage would it be useful/possible to use this library http://www.baeldung.com/spring-cloud-sleuth-single-application or at least re-use some design ideas?

Reply:
Think that looks like it's pretty closely tied to Spring.  I'm guessing it works in with Spring's request handling. Certainly some of the examples on that page use Spring wiring.